firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java

```diff
@@ -0,0 +1,240 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.core;
+
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import androidx.annotation.Nullable;
+import com.google.firebase.firestore.core.OrderBy.Direction;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.ResourcePath;
+import java.util.List;
+
+/**
+ * A Target represents the WatchTarget representation of a Query, which is used by the LocalStore
+ * and the RemoteStore to keep track of and to execute backend queries. While multiple Query can map
+ * to the same Target, each Targets maps to a single WatchTarget in RemoteStore and a single
+ * TargetData entry in persistence.
+ */
+public final class Target {
+  public static final long NO_LIMIT = -1;
+
+  private @Nullable String memorizedCannonicalId;
+
+  private final List<OrderBy> explicitSortOrder;
+  private final List<Filter> filters;
+
+  private final ResourcePath path;
+
+  private final @Nullable String collectionGroup;
+
+  private final long limit;
+
+  private final @Nullable Bound startAt;
+  private final @Nullable Bound endAt;
+
+  /**
+   * Initializes a Target with a path and additional query constraints. Path must currently be empty
+   * if this is a collection group query.
+   *
+   * <p>NOTE: you should always construct Target from {@code Query.toTarget} instead of using this
+   * constructor, because Query provides an implicit {@code orderBy} property.
```

Comment:
I left a related comment below, but this seems a little weird to me.  The reason Query has the implicit vs explicit thing is because we let users omit some of the necessary orderBy clauses and then we implicitly add them (e.g. whenever you do an inequality filter you must also order by that field).  But when we construct a Target directly, we should probably be creating it with the correct orderBy's directly.

I don't feel super strongly though.  If we think this is safer / cleaner, I'm fine with it.

Reply:
When I wrote this in JS, I was not thinking about the case of de-serialization. My plan is to make `Target` constructor only callable from the same package..but TS does not support package-private access?

I forgot to port this here though..now I made `Target` constructor `default` access. It does make de-serialization a little awkward, but now at least people would not be able to construct a `Target` directly. I think this is slightly better.